signer/src/network/mod.rs

```diff
@@ -34,6 +36,10 @@ pub trait MessageTransfer: Clone {
     fn broadcast(&mut self, msg: Msg) -> impl Future<Output = Result<(), Error>> + Send;
     /// Receive a message from the network
     fn receive(&mut self) -> impl Future<Output = Result<Msg, Error>> + Send;
+    /// Return a stream of the same messages from the
+    /// [`MessageTransfer::receive`]. The inner type of each
+    /// [`SignerSignal`] should be a [`P2PEvent::MessageReceived`].
+    fn receiver_stream(&self) -> BroadcastStream<SignerSignal>;
```

Comment:
Did this need to explicitly be a `BroadcastStream` or would an `impl Stream` work?

Also it feels a little weird that this trait is returning a stream of `SignerSignal`s -- but maybe you're thinking in the lines of on-the-road-to-killing-this-trait?

Reply:
> Also it feels a little weird that this trait is returning a stream of `SignerSignal`s -- but maybe you're thinking in the lines of on-the-road-to-killing-this-trait?

Returning a `Msg` makes the most sense, and that's what I tried first. But it made things more complex in the `new_signal_stream` implementation.

Although I thought about it killing it, the trait seems very useful, I'm not sure if we can or should.

Reply:
Yeah I would have expected `Msg`, but I also get why you did it this way. I think we should revisit this later because it feels "wrong", which I think the comment needing to specify the valid variants kinda implicitly indicates.

Reply:
Yeah, I agree, it does seem wrong.